### PR TITLE
[FLINK-2678]DataSet API does not support multi-dimensional arrays as keys

### PR DESCRIPTION
**Pull request by sbcd90**
Hello,

@tillrohrmann I have added support for multi-dimensional arrays as keys in Dataset api. Please review & merge.

### CODE REVIEW

**Comment by sbcd90** [comment]
@StephanEwen could you also help reviewing the code?

File: flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.core.memory.MemorySegment;
+
+import java.io.IOException;
+import java.lang.reflect.Array;
+import java.util.Arrays;
+
+public class GenericArrayComparator<T> extends TypeComparator<T[]> implements java.io.Serializable {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java]: Java docs missing

File: flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	public boolean invertNormalizedKey() {
+		return !ascendingComparison;
+	}
+
+	@Override
+	public int hash(T[] record) {
+		return Arrays.hashCode(record);
+	}
+
+	private int compareValues(Object first, Object second) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java]: This only works for primitive types. What about complex types?

File: flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+			}
+		}
+		else {
+			compareResult = compareValues(firstArray, secondArray);
+		}
+		return compareResult;
+	}
+
+	@Override
+	public int compare(T[] first, T[] second) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java]: I think the compare method should only check that the length of `first` and `second` are the same and then apply the component comparator (derived from `ObjectArrayTypeInfo.componentInfo`) on the elements.

File: flink-core/src/test/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparatorTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+package org.apache.flink.api.common.typeutils.base;
+
+import org.apache.flink.api.common.typeinfo.PrimitiveArrayTypeInfo;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.ComparatorTestBase;
+import org.apache.flink.api.common.typeutils.TypeComparator;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.junit.Assert;
+
+public class GenericArrayComparatorTest extends ComparatorTestBase<char[][]> {
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparatorTest.java]: Non primitive types should be tested as well.

File: flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java]: Why not calling it `ObjectArrayComparator`?

File: flink-java/src/test/java/org/apache/flink/api/java/operator/DataSinkTest.java [hunk 5]
```diff
@@ -311,7 +311,7 @@ public void testPojoSingleOrderFull() {
 			.sortLocalOutput("*", Order.ASCENDING);
 	}
 
-	@Test(expected = InvalidProgramException.class)
+	@Test
 	public void testArrayOrderFull() {
```
Comment (tillrohrmann) [inline flink-java/src/test/java/org/apache/flink/api/java/operator/DataSinkTest.java]: That is not how you fix a test. When the behaviour of the program changes, then one should also adapt the test accordingly. E.g. checking that the order of the elements is correct.

**Comment by tillrohrmann** [comment]
Thanks for your contribution @sbcd90. I think the PR needs still a bit of work before it can be merged. The implementation of `GenericArrayComparator` only works for primitive types. Furthermore, I think that it is easier to construct the comparator out of the component type information of the `ObjectArrayTypeInfo` instead of doing the type analysis again every time you call `compare`.

File: flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java (hunk 3 above)
Comment (sbcd90) [inline flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java]: `ObjectArrayTypeInfo.componentInfo` is of type `TypeInformation<T>` & accessing a method like `createComparator` from one of the implementing classes of `TypeInformation<T>` is difficult. One possible solution is to have a static map like `TYPES` which maps classes to comparator classes. From `ObjectArrayTypeInfo.componentInfo` I get the typeClass & access the corresponding comparator from the map. Is there any other better solution?

File: flink-core/src/test/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparatorTest.java (hunk 4 above)
Comment (sbcd90) [inline flink-core/src/test/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparatorTest.java]: Should another test class be created for testing non-primitive types?

**Comment by sbcd90** [comment]
@tillrohrmann Thanks for the review. I had a few questions based on your comments. Kindly help me in getting the questions answered so that I can proceed.

File: flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java (hunk 3 above)
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java]: You should check whether `componentInfo` is an `AtomicType` or a `ComponentType` and then create the comparators via the provided methods. You can take a look at the `ComponentType.createComparator` implementation.

**Comment by sbcd90** [comment]
Hello @tillrohrmann  ..I have made all the changes you have mentioned.

- support for CompositeType.
- added test cases for both primitive type & composite type
- modified the changed test case in DataSinkTest.java
- changed the name of Comparator to ObjectArrayComparator.java
- javadocs are auto-generated by intellij.

Please please please please review my code now. Kindly let me know if there are any further changes required. Kindly let me know if this can be merged now also.

**Comment by sbcd90** [comment]
Hello @tillrohrmann please review the commit as the work to be done is nearly complete I believe. Please comment.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java [hunk 6]
```diff
@@ -72,15 +77,59 @@ public int getTotalFields() {
...
 	public TypeSerializer<T> createSerializer(ExecutionConfig executionConfig) {
 		return (TypeSerializer<T>) new GenericArraySerializer<C>(
-			componentInfo.getTypeClass(),
-			componentInfo.createSerializer(executionConfig));
+				componentInfo.getTypeClass(),
+				componentInfo.createSerializer(executionConfig));
+	}
+
+	@SuppressWarnings("unchecked")
+	private TypeComparator<? super Object> getBaseComparatorInfo(TypeInformation<? extends Object> componentInfo, boolean sortOrderAscending, ExecutionConfig executionConfig) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java]: Why are you extracting for certain types the element comparator (for example the ObjectArrayTypeInfo) and for other you create the `TypeComparator` (for example the `CompositeTypeInfo`)? I don't get it. Why do you need the `getBaseComparatorInfo` method at all? Simply check in `createComparator` the different subtypes and then create the `TypeComparator`.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java [hunk 7]
```diff
@@ -72,15 +77,59 @@ public int getTotalFields() {
...
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public TypeComparator<T> createComparator(boolean sortOrderAscending, ExecutionConfig executionConfig) {
+
+		return (TypeComparator<T>) new ObjectArrayComparator<T,C>(
+			sortOrderAscending,
+			(GenericArraySerializer<T>) createSerializer(executionConfig),
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java]: Why this cast here?

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/ObjectArrayComparator.java [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+	private int compareValues(Object first, Object second) {
+		/**
+		 * uses the chosen comparator ( of primitive or composite type ) & compares the provided objects as input
+		 */
+		return comparatorInfo.compare(first, second);
+	}
+
+	@SuppressWarnings("unchecked")
+	private int parseGenericArray(Object firstArray, Object secondArray) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/ObjectArrayComparator.java]: Why exactly are you doing all the array operation on `Objects`?

Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/ObjectArrayComparator.java]: I think this is not the right way to go. Simply check then length of the arrays in the `compare` method and then call for all the elements the type comparator of the element type of the array.

File: flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/ObjectArrayComparatorCompositeTypeTest.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Tuple2<String, Integer>[][][] getSortedTestData() {
+		Object result = Array.newInstance(Tuple2.class, new int[]{2, 2, 1});
+
+		((Tuple2<String, Integer>[][][]) result)[0][0][0] = new Tuple2<String, Integer>();
+		((Tuple2<String, Integer>[][][]) result)[0][0][0].f0 = "be";
+		((Tuple2<String, Integer>[][][]) result)[0][0][0].f1 = 2;
+
+		((Tuple2<String, Integer>[][][]) result)[0][1][0] = new Tuple2<String, Integer>();
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/ObjectArrayComparatorCompositeTypeTest.java]: you can write something like `Tuple2.of("not", 3)`. Makes your life easier.

**Comment by tillrohrmann** [comment]
I think the PR is still not in a good shape. I pointed out the problematic code sections.

Also please bear in mind that people will always try to review your code as soon as possible. But sometimes people are just busy and it takes some time for them to review a PR. In such a case, one simply has to be a little bit patient with them.

**Pull request closed without merging**